conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureBodySerDe.java

```diff
@@ -0,0 +1,179 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+/** Package private internal API. */
+final class ConjureBodySerDe implements BodySerDe {
+
+    private static final String BINARY_CONTENT_TYPE = "application/octet-stream";
+
+    private final List<Encoding> encodings;
+
+    /**
+     * Selects the first (based on input order) of the provided encodings that
+     * {@link Encoding#supportsContentType supports} the serialization format {@link Headers#ACCEPT accepted}
+     * by a given request, or the first serializer if no such serializer can be found.
+     */
+    ConjureBodySerDe(List<Encoding> encodings) {
+        // Defensive copy
+        this.encodings = ImmutableList.copyOf(encodings);
+        Preconditions.checkArgument(encodings.size() > 0, "At least one Encoding is required");
+    }
+
+    @Override
+    public <T> Serializer<T> serializer(TypeToken<T> token) {
+        return new EncodingSerializerRegistry<>(encodings, token);
+    }
+
+    @Override
+    public <T> Deserializer<T> deserializer(TypeToken<T> token) {
+        return new EncodingDeserializerRegistry<>(encodings, token);
+    }
+
+    @Override
+    public void serialize(BinaryResponseBody value, HttpServerExchange exchange) throws IOException {
+        Preconditions.checkNotNull(value, "A BinaryResponseBody value is required");
+        exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, BINARY_CONTENT_TYPE);
+        value.write(exchange.getOutputStream());
+    }
+
+    @Override
+    public InputStream deserializeInputStream(HttpServerExchange exchange) {
+        String contentType = exchange.getRequestHeaders().getFirst(Headers.CONTENT_TYPE);
+        if (contentType == null) {
+            throw new SafeIllegalArgumentException("Request is missing Content-Type header");
+        }
+        if (!contentType.startsWith(BINARY_CONTENT_TYPE)) {
+            throw new SafeIllegalArgumentException("Unsupported Content-Type",
+                    SafeArg.of("Content-Type", contentType));
+        }
+        return exchange.getInputStream();
+    }
+
+    private static final class EncodingSerializerRegistry<T> implements Serializer<T> {
+
+        private final EncodingSerializerContainer<T> defaultEncoding;
+        private final List<EncodingSerializerContainer<T>> encodings;
+
+        EncodingSerializerRegistry(List<Encoding> encodings, TypeToken<T> token) {
+            this.encodings = encodings.stream()
+                    .map(encoding -> new EncodingSerializerContainer<>(encoding, token))
+                    .collect(ImmutableList.toImmutableList());
+            this.defaultEncoding = this.encodings.get(0);
+        }
+
+        @Override
+        public void serialize(T value, HttpServerExchange exchange) throws IOException {
+            Preconditions.checkNotNull(value, "cannot serialize null value");
+            EncodingSerializerContainer<T> container = getResponseSerializer(exchange);
```

Comment:
tiny bit subtle that the second param of `Serializer#serialize(T, HttpServerExchange)` is only intended for people to _write_ data, but here we're _reading_ the original request's content-type and changing behaviour based on it. 

Don't have a better proposal here - just something that made me look twice